### PR TITLE
Allow user control over pod labels using controllerconfig

### LINKED ISSUE #2398: Add static label to providers identifying the package
### What problem are you facing?
We have installed `provider-sql` using the `Provider` CRD. I would like to restart the deployment using a predictable name, and add security groups to the pods, but the names of the pods/deployments are generated. One way to achieve this would be to use label selectors.

```
apiVersion: pkg.crossplane.io/v1beta1
kind: Provider
metadata:
  name: crossplane-provider-sql
spec:
  package: crossplane/provider-sql:v0.2.0
```
```
    labels:
      pkg.crossplane.io/revision: crossplane-provider-sql-16bdf1af9403
      pod-template-hash: 997d559fb
```

### How could Crossplane help solve your problem?
Add a label at least for the package name, without tags. It might be useful to add more labels too as they will show up in prometheus etc.

### PR DESCRIPTION
**Pull request by chlunde**
To be able to select pods using PodSecurityPolicy in EKS, we need to
have predictable labels on the pods. Currently only an unpredictable
hash label is provided on the pods, and the user can only add
annotations.

This change adds a label map for the pod template, similar to the
existing annotation map.

To ensure future extensibility labels with crossplane.io prefixes are
forbidden in the documentation, and the code is structured to ensure
that the template will match the pod selector, and that any user
provided labels which overlaps with the selector will be overriden.

Fixes #2398

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested
* existing tests work, but coverage is low for ControllerConfig
* checked with kind cluster

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by chlunde** [comment]
```yaml
apiVersion: pkg.crossplane.io/v1alpha1
kind: ControllerConfig
metadata:
  name: sql-config
  # https://github.com/crossplane/crossplane/issues/2398#issuecomment-867570603
  labels:
    provider: sql
spec:
  metadata:
    labels:
      foo: bar
```

```yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Pod
  metadata:
    creationTimestamp: "2021-08-03T11:12:16Z"
    generateName: crossplane-provider-sql-16bdf1af9403-546677cfb5-
    labels:
      foo: bar
      pkg.crossplane.io/revision: crossplane-provider-sql-16bdf1af9403
      pod-template-hash: 546677cfb5
```

**Comment by negz** [comment]
@turkenh @ulucinar would one of you mind owning review for this PR? I'm happy to merge if and when it looks good to you. Thank you!

**Comment by chlunde** [comment]
I would also be happy with two alternative implementations:
* just add another fixed label with the package name
* copy the labels from the controllerconfig

File: apis/pkg/v1alpha1/config.go
Comment (hasheddan) [inline apis/pkg/v1alpha1/config.go:150]: nitpick
```suggestion
	// with internal labels used by crossplane, and labels with a
	// crossplane.io key might be overwritten.
```

**Comment by chlunde** [comment]
@turkenh can you trigger CI run here? "First-time contributors need a maintainer to approve running workflows"

**Comment by turkenh** [comment]
> @turkenh can you trigger CI run here? "First-time contributors need a maintainer to approve running workflows"

@chlunde I am afraid I don't have that right yet. Maybe @muvaf can help?

File: internal/controller/pkg/revision/deployment.go
Comment (turkenh) [inline internal/controller/pkg/revision/deployment.go:107]: How would the template matches the selector if no controller config provided?

File: internal/controller/pkg/revision/deployment.go
Comment (turkenh) [inline internal/controller/pkg/revision/deployment.go:66]: Similar/related to the other comment, we would lose this label if no controller config is provided.

File: internal/controller/pkg/revision/deployment.go
Comment (chlunde) [reply internal/controller/pkg/revision/deployment.go]: Thanks! Fixed

File: internal/controller/pkg/revision/deployment_test.go
Comment (negz) [inline internal/controller/pkg/revision/deployment_test.go:1]: Thank you for adding these tests @chlunde!

File: internal/controller/pkg/revision/deployment_test.go
Comment (negz) [inline internal/controller/pkg/revision/deployment_test.go:15]: The linter comment is a little vague here; it's asking you to separate out all the `crossplane/crossplane` (i.e. "local") imports into a group separated by a line of whitespace. We typically separate out runtime imports too, so it looks something like:

```go
import (
	"stdlib-imports"

	"third-party-imports"

	"crossplane-runtime-imports"

	"crossplane-imports"
)

Comment (chlunde) [reply internal/controller/pkg/revision/deployment_test.go:15]: :+1: 

forgot `make reviewable` after adding the test.

Running `goimports -w -local github.com/crossplane/crossplane internal/controller/pkg/revision/deployment_test.go` now

File: internal/controller/pkg/revision/deployment_test.go
Comment (chlunde) [reply internal/controller/pkg/revision/deployment_test.go:1]: Lesson learned, even for a "trivial" change, add tests even if the file does not have any already :)

PS! This doesn't add full coverage for all the existing branches here

**Review by ulucinar — APPROVED** [review]
LGTM. Thank you @chlunde!

File: internal/controller/pkg/revision/deployment_test.go
Comment (chlunde) [reply internal/controller/pkg/revision/deployment_test.go:13]: right, running `goimports -w -local github.com/crossplane/crossplane internal/controller/pkg/revision/deployment_test.go` now wasn't enough...

File: internal/controller/pkg/revision/deployment_test.go
Comment (hasheddan) [inline internal/controller/pkg/revision/deployment_test.go:1]: @chlunde thanks again for adding tests here :) could you make sure to add the appropriate license header to this file?

**Review by hasheddan — APPROVED** [review]

**Pull request merged**
